core/src/main/scala/akka/kafka/internal/ExternalSingleSourceLogic.scala

```diff
@@ -0,0 +1,100 @@
+import akka.kafka.Subscription.{Assignment, AssignmentWithOffset, TopicSubscription, TopicSubscriptionPattern}
+import akka.kafka.{KafkaConsumerActor, Subscription}
+import akka.stream.stage.GraphStageLogic.StageActor
+import akka.stream.stage.{GraphStageLogic, OutHandler}
+import akka.stream.{ActorMaterializer, SourceShape}
+import org.apache.kafka.clients.consumer.ConsumerRecord
+import org.apache.kafka.common.TopicPartition
+
+import scala.annotation.tailrec
+
+private[kafka] abstract class ExternalSingleSourceLogic[K, V, Msg](
+    val shape: SourceShape[Msg],
+    consumer: ActorRef,
+    subscription: Subscription
+) extends GraphStageLogic(shape) with PromiseControl with MessageBuilder[K, V, Msg] {
+  var tps = Set.empty[TopicPartition]
+  var buffer: Iterator[ConsumerRecord[K, V]] = Iterator.empty
+  var requested = false
+  var self: StageActor = _
+  implicit lazy val ec = ActorMaterializer.downcast(materializer).executionContext
+
+  override def preStart(): Unit = {
+    super.preStart()
+    setKeepGoing(true)
+    subscription match {
+      case TopicSubscription(topics) =>
+        consumer ! KafkaConsumerActor.Subscribe(topics, KafkaConsumerActor.rebalanceListener(partitionAssignedCB.invoke, partitionRevokedCB.invoke))
+      case TopicSubscriptionPattern(topics) =>
+        consumer ! KafkaConsumerActor.SubscribePattern(topics, KafkaConsumerActor.rebalanceListener(partitionAssignedCB.invoke, partitionRevokedCB.invoke))
+      case Assignment(topics) =>
+        consumer ! KafkaConsumerActor.Assign(topics)
+        tps ++= topics
+      case AssignmentWithOffset(topics) =>
+        consumer ! KafkaConsumerActor.AssignWithOffset(topics)
+        tps ++= topics.keySet
+    }
+
+    self = getStageActor {
+      case (sender, msg: KafkaConsumerActor.Messages[K, V]) =>
+        requested = false
+        // do not use simple ++ because of https://issues.scala-lang.org/browse/SI-9766
+        if (buffer.hasNext) {
+          buffer = buffer ++ msg.messages
+        }
+        else {
+          buffer = msg.messages
+        }
+        pump()
+      case (_, Terminated(ref)) if ref == consumer =>
+        failStage(new Exception("Consumer actor terminated"))
+    }
+  }
+
+  val partitionAssignedCB = getAsyncCallback[Iterable[TopicPartition]] { newTps =>
+    tps ++= newTps
+    pump()
+  }
+  val partitionRevokedCB = getAsyncCallback[Iterable[TopicPartition]] { newTps =>
+    tps --= newTps
+    pump()
+  }
+
+  @tailrec
+  private def pump(): Unit = {
+    if (isAvailable(shape.out)) {
+      if (buffer.hasNext) {
+        val msg = buffer.next()
+        push(shape.out, createMessage(msg))
+        pump()
+      }
+      else if (!requested && tps.nonEmpty) {
+        requested = true
+        consumer.tell(RequestMessages(tps), self.ref)
+      }
+    }
+  }
+
+  setHandler(shape.out, new OutHandler {
+    override def onPull(): Unit = {
+      pump()
+    }
+  })
+
+  override def performShutdown() = {
```

Comment:
Also, I don't see this called. If you set keep going flag, then you *must* explicitly call completeStage() even if all ports are closed, otherwise you leak the stage.